File: python/parallel.py
import math
import time
import numpy as np
from multiprocessing import Pool, cpu_count
from sequential import generate_fractal_tree
from utils import print_header, print_params, print_result


def _worker(args):
    x, y, length, angle, ratio, branch_angle_rad, min_length, depth = args
    branches = generate_fractal_tree(x, y, length, angle, ratio, branch_angle_rad, min_length, start_depth=depth)
    return np.array(branches, dtype=np.float64)


def _build_tasks(x, y, length, angle, ratio, branch_angle_rad, min_length, depth, target_depth, upper_branches):
    if depth >= target_depth:
        return [(x, y, length, angle, ratio, branch_angle_rad, min_length, depth)]

    end_x = x + length * math.cos(angle)
    end_y = y + length * math.sin(angle)
    upper_branches.append((x, y, end_x, end_y, depth))

    new_len = length * ratio
    left = _build_tasks(end_x, end_y, new_len, angle + branch_angle_rad, ratio, branch_angle_rad, min_length, depth + 1, target_depth, upper_branches)
    right = _build_tasks(end_x, end_y, new_len, angle - branch_angle_rad, ratio, branch_angle_rad, min_length, depth + 1, target_depth, upper_branches)
    return left + right


def run_parallel(trunk_length=100.0, ratio=0.67, branch_angle=30.0,
                        min_length=0.01, num_processes=None):

    if num_processes is None:
        num_processes = cpu_count()

    branch_angle_rad = math.radians(branch_angle)
    split_depth = max(1, math.ceil(math.log2(num_processes * 4)))

    print_header("Parallel (Python)")
    print_params(trunk_length, ratio, branch_angle, min_length,
                 cores=num_processes, split_depth=split_depth)

    start_time = time.perf_counter()

    # Build upper levels sequentially
    start_x, start_y = 0, 0
    start_angle = math.pi / 2
    end_x = start_x + trunk_length * math.cos(start_angle)
    end_y = start_y + trunk_length * math.sin(start_angle)
    start_depth = 0

    upper_branches = [(start_x, start_y, end_x, end_y, start_depth)]
    new_len = trunk_length * ratio
    left_child_angle_rad = start_angle + branch_angle_rad
    right_child_angle_rad = start_angle - branch_angle_rad

    # left subtree
    tasks = _build_tasks(end_x, end_y, 
                         new_len, left_child_angle_rad,
                         ratio, branch_angle_rad, 
                         min_length, 1, 
                         split_depth, upper_branches)
    
    # right subtree
    tasks += _build_tasks(end_x, end_y, 
                          new_len, right_child_angle_rad,
                          ratio, branch_angle_rad, 
                          min_length, 1, 
                          split_depth, upper_branches)

    with Pool(processes=num_processes) as pool:
        results = pool.map(_worker, tasks)

    upper_array = np.array(upper_branches, dtype=np.float64)
    branches = np.concatenate([upper_array] + results)

    execution_time = time.perf_counter() - start_time

    total_branches = len(branches)
    max_depth = int(branches[:, 4].max()) if total_branches > 0 else 0
    print_result(execution_time, total_branches, max_depth)

    result = {
        'parameters': {
            'trunk_length': trunk_length,
            'ratio': ratio,
            'branch_angle': branch_angle,
            'min_length': min_length,
            'split_depth': split_depth,
            'num_processes': num_processes,
        },
        'execution_time': execution_time,
    }

    return result


if __name__ == "__main__":

    run_parallel(
        trunk_length=100.0,
        ratio=0.67,
        branch_angle=30.0,
        min_length=0.01,
    )

File: rust/src/parallel.rs
use fractal_tree::{
    generate_fractal_tree, collect_tasks,
    print_header, print_params, print_extra_param, print_result, save_result,
    Branch, FractalResult, Parameters,
};
use rayon::prelude::*;
use std::f64::consts::PI;
use std::thread;
use std::time::Instant;
use std::error::Error;


pub fn run_parallel(
    trunk_length: f64, 
    ratio: f64, 
    branch_angle: f64, 
    min_length: f64,
    num_threads: Option<usize>, 
    output_file: &str,
) -> Result<FractalResult, Box<dyn Error>> {

    let num_threads = num_threads.unwrap_or_else(|| {
        thread::available_parallelism().map(|n| n.get()).unwrap_or(4)
    });

    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(num_threads)
        .build()?;

    let branch_angle_rad = branch_angle.to_radians();
    let split_depth = (num_threads * 4).next_power_of_two().ilog2() as usize;

    print_header("Parallel Simple (Rust)");
    print_params(trunk_length, ratio, branch_angle, min_length);
    print_extra_param("threads", &num_threads);
    print_extra_param("split_depth", &split_depth);

    let start = Instant::now();

    let mut upper_branches = vec![Branch { x1: 0.0, y1: 0.0, x2: 0.0, y2: trunk_length, depth: 0 }];
    let child_length = trunk_length * ratio;
    let mut tasks = collect_tasks(&mut upper_branches, 0.0, trunk_length, child_length, PI / 2.0 + branch_angle_rad, ratio, branch_angle_rad, 1, split_depth);
    tasks.extend(collect_tasks(&mut upper_branches, 0.0, trunk_length, child_length, PI / 2.0 - branch_angle_rad, ratio, branch_angle_rad, 1, split_depth));

    let subtree_results: Vec<Vec<Branch>> = pool.install(|| {
        tasks.par_iter()
            .map(|task| generate_fractal_tree(task.x, task.y, task.length, task.angle, ratio, branch_angle_rad, min_length, task.depth))
            .collect()
    });

    let execution_time = start.elapsed().as_secs_f64();

    let total_branches: usize = upper_branches.len() + subtree_results.iter().map(|subtree| subtree.len()).sum::<usize>();
    let max_depth = subtree_results.iter()
        .filter_map(|subtree| subtree.last().map(|branch| branch.depth))
        .max()
        .unwrap_or(0);

    print_result(execution_time, total_branches, max_depth);

    let mut result = FractalResult {
        parameters: Parameters { trunk_length, ratio, branch_angle, min_length },
        execution_time,
        total_branches,
        max_depth,
        iterations: Vec::new(),
    };

    let mut all_groups = vec![upper_branches];
    all_groups.extend(subtree_results);
    save_result(&mut result, &all_groups, output_file)?;

    Ok(result)
}

fn main() -> Result<(), Box<dyn Error>> {
    run_parallel(100.0, 0.67, 30.0, 0.01, None, "parallel_rust.json")?;
    Ok(())
}
